lib/ansible/modules/cloud/amazon/ec2_group.py

```diff
@@ -645,21 +579,181 @@ def fix_port_and_protocol(permission):
+    return changed
+
+
+def check_rule_description_update(client, module, rule_id, rule, grant_type, current_rules):
+    needs_update = False
     if 'rule_desc' in rule:
-        current_rule_description = rule.get('rule_desc') or ''
-        if current_rule_description != groupRules[rule_id][0]['UserIdGroupPairs'][0].get('Description', ''):
-            if not module.check_mode:
-                ip_permission = serialize_group_grant(group_id, rule)
-                update_rules_description(module, client, rule_type, group['GroupId'], ip_permission)
-            changed = True
+        new_rule_description = rule.get('rule_desc') or ''
+        if grant_type == 'group':
+            current_rule_description = current_rules[rule_id][0]['UserIdGroupPairs'][0].get('Description', '')
+        else:
+            current_rule_description = (
+                current_rules[rule_id][0].get('IpRanges')
+                or current_rules[rule_id][0].get('Ipv6Ranges')
+            )[0].get('Description')
+        if new_rule_description != current_rule_description:
+            needs_update = True
+    return needs_update
+
+
+def update_tags(client, module, group_id, current_tags, tags, purge_tags, changed):
+    tags_need_modify, tags_to_delete = compare_aws_tags(current_tags, tags, purge_tags)
+    changed |= bool(tags_need_modify or tags_to_delete)
+
+    if not module.check_mode:
+        if tags_to_delete:
+            try:
+                client.delete_tags(Resources=[group_id], Tags=[{'Key': tag} for tag in tags_to_delete])
+            except (BotoCoreError, ClientError) as e:
+                module.fail_json_aws(e, msg="Unable to delete tags {0}".format(tags_to_delete))
+
+        # Add/update tags
+        if tags_need_modify:
+            try:
+                client.create_tags(Resources=[group_id], Tags=ansible_dict_to_boto3_tag_list(tags_need_modify))
+            except (BotoCoreError, ClientError) as e:
+                module.fail_json(e, msg="Unable to add tags {0}".format(tags_need_modify))
+
     return changed
 
 
+def create_security_group(client, module, name, description, vpc_id):
+    if not module.check_mode:
+        params = dict(GroupName=name, Description=description)
+        if vpc_id:
+            params['VpcId'] = vpc_id
+        try:
+            group = client.create_security_group(**params)
+        except (BotoCoreError, ClientError) as e:
+            module.fail_json_aws(e, msg="Unable to create security group")
+        # When a group is created, an egress_rule ALLOW ALL
+        # to 0.0.0.0/0 is added automatically but it's not
+        # reflected in the object returned by the AWS API
+        # call. We re-read the group for getting an updated object
+        # amazon sometimes takes a couple seconds to update the security group so wait till it exists
+        while True:
+            group = get_security_groups_with_backoff(client, GroupIds=[group['GroupId']])['SecurityGroups'][0]
```

Comment:
I worry that this will *cause* throttling as this is going to call repeatedly with no backoff until throttling occurs.

Might be worth adding a short sleep before each retry at the very least.